In [ ]:
! pip install pandas
! pip install numPy
! pip install seaborn
! pip install matplotlib

In [4]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [7]:
data=pd.read_excel("./Data.xlsx")
rawData=data.copy()
rawData.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
...,...,...,...,...,...,...,...,...
541904,581587,22613,PACK OF 20 SPACEBOY NAPKINS,12,2011-12-09 12:50:00,0.85,12680.0,France
541905,581587,22899,CHILDREN'S APRON DOLLY GIRL,6,2011-12-09 12:50:00,2.10,12680.0,France
541906,581587,23254,CHILDRENS CUTLERY DOLLY GIRL,4,2011-12-09 12:50:00,4.15,12680.0,France
541907,581587,23255,CHILDRENS CUTLERY CIRCUS PARADE,4,2011-12-09 12:50:00,4.15,12680.0,France


Know Your Data

In [12]:
rawData.tail(10)
rawData.shape
rawData.columns
rawData.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[ns]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      541909 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 33.1+ MB


In [13]:
rawData.describe()

,Quantity,InvoiceDate,UnitPrice,CustomerID
count,541909.000000,541909,541909.000000,406829.000000
mean,9.552250,2011-07-04 13:34:57.156386048,4.611114,15287.690570
min,-80995.000000,2010-12-01 08:26:00,-11062.060000,12346.000000
25%,1.000000,2011-03-28 11:34:00,1.250000,13953.000000
50%,3.000000,2011-07-19 17:17:00,2.080000,15152.000000
75%,10.000000,2011-10-19 11:27:00,4.130000,16791.000000
max,80995.000000,2011-12-09 12:50:00,38970.000000,18287.000000
std,218.081158,NaN,96.759853,1713.600303


In [15]:
rawData.dtypes

InvoiceNo              object
StockCode              object
Description            object
Quantity                int64
InvoiceDate    datetime64[ns]
UnitPrice             float64
CustomerID            float64
Country                object
dtype: object

In [18]:
rawData["InvoiceNo"].nunique()

25900

In [20]:
rawData["Country"].nunique()


38

In [22]:
rawData["InvoiceDate"].min()
rawData["InvoiceDate"].max()

Timestamp('2011-12-09 12:50:00')

In [23]:
print("Unique Invoices:", rawData["InvoiceNo"].nunique())
print("Unique Products:", rawData["StockCode"].nunique())
print("Unique Customers:", rawData["CustomerID"].nunique())
print("Unique Countries:", rawData["Country"].nunique())

Unique Invoices: 25900
Unique Products: 4070
Unique Customers: 4372
Unique Countries: 38


In [24]:
rawData.sample()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
62619,541508,47566b,TEA TIME PARTY BUNTING,1,2011-01-18 16:06:00,8.29,NaN,United Kingdom


1.Handling Missing Values:

In [25]:
missing_count = rawData.isnull().sum()

missing_percentage = (missing_count / len(rawData)) * 100

missing_summary = pd.DataFrame({
    "Missing Values": missing_count,
    "Missing Percentage": missing_percentage
})

missing_summary

,Missing Values,Missing Percentage
InvoiceNo,0,0.000000
StockCode,0,0.000000
Description,1454,0.268311
Quantity,0,0.000000
InvoiceDate,0,0.000000
UnitPrice,0,0.000000
CustomerID,135080,24.926694
Country,0,0.000000


In [26]:
rawData.isnull().sum()[rawData.isnull().sum() > 0]

Description      1454
CustomerID     135080
dtype: int64

In [30]:
missing_desc = rawData[rawData["Description"].isnull()]

missing_desc[["StockCode", "Description", "Quantity", "UnitPrice", "Country"]].head(20)


,StockCode,Description,Quantity,UnitPrice,Country
622,22139,NaN,56,0.0,United Kingdom
1970,21134,NaN,1,0.0,United Kingdom
1971,22145,NaN,1,0.0,United Kingdom
1972,37509,NaN,1,0.0,United Kingdom
1987,85226A,NaN,1,0.0,United Kingdom
1988,85044,NaN,1,0.0,United Kingdom
2024,20950,NaN,1,0.0,United Kingdom
2025,37461,NaN,3,0.0,United Kingdom
2026,84670,NaN,23,0.0,United Kingdom
2406,21777,NaN,-10,0.0,United Kingdom


In [31]:
missing_desc["StockCode"].nunique()

960

In [32]:
description_check = rawData[
    rawData["StockCode"].isin(missing_desc["StockCode"]) &
    rawData["Description"].notnull()
]

description_check[["StockCode", "Description"]].drop_duplicates().head(20)

,StockCode,Description
2,84406B,CREAM CUPID HEARTS COAT HANGER
3,84029G,KNITTED UNION FLAG HOT WATER BOTTLE
4,84029E,RED WOOLLY HOTTIE WHITE HEART.
7,22633,HAND WARMER UNION JACK
8,22632,HAND WARMER RED POLKA DOT
19,21777,RECIPE BOX WITH METAL HEART
28,22726,ALARM CLOCK BAKELIKE GREEN
31,10002,INFLATABLE POLITICAL GLOBE
45,POST,POSTAGE
53,37370,RETRO COFFEE MUGS ASSORTED


In [33]:
missing_description = rawData[rawData["Description"].isna()]

missing_description[["StockCode", "Description"]].head(20)

,StockCode,Description
622,22139,NaN
1970,21134,NaN
1971,22145,NaN
1972,37509,NaN
1987,85226A,NaN
1988,85044,NaN
2024,20950,NaN
2025,37461,NaN
2026,84670,NaN
2406,21777,NaN


In [34]:
description_mapping = (
    rawData.dropna(subset=["Description"])
           .drop_duplicates("StockCode")
           .set_index("StockCode")["Description"]
)

description_mapping.head()

StockCode
85123A     WHITE HANGING HEART T-LIGHT HOLDER
71053                     WHITE METAL LANTERN
84406B         CREAM CUPID HEARTS COAT HANGER
84029G    KNITTED UNION FLAG HOT WATER BOTTLE
84029E         RED WOOLLY HOTTIE WHITE HEART.
Name: Description, dtype: object

In [35]:
rawData["Description"] = rawData["Description"].fillna(
    rawData["StockCode"].map(description_mapping)
)

In [39]:
rawData.isnull().sum()

InvoiceNo           0
StockCode           0
Description       112
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

There were initially 1,454 missing values in the Description column (0.27%). I used the StockCode to find matching product descriptions from other records and successfully recovered 1,342 values. The remaining 112 values could not be reliably identified, so I kept them as missing instead of guessing or adding incorrect information. These records were retained because other transaction information such as quantity, price, date, and country is still available.

In [40]:
missing_customer = rawData[rawData["CustomerID"].isna()]

missing_customer.head(10)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
622,536414,22139,RETROSPOT TEA SET CERAMIC 11 PC,56,2010-12-01 11:52:00,0.00,NaN,United Kingdom
1443,536544,21773,DECORATIVE ROSE BATHROOM BOTTLE,1,2010-12-01 14:32:00,2.51,NaN,United Kingdom
1444,536544,21774,DECORATIVE CATS BATHROOM BOTTLE,2,2010-12-01 14:32:00,2.51,NaN,United Kingdom
1445,536544,21786,POLKADOT RAIN HAT,4,2010-12-01 14:32:00,0.85,NaN,United Kingdom
1446,536544,21787,RAIN PONCHO RETROSPOT,2,2010-12-01 14:32:00,1.66,NaN,United Kingdom
1447,536544,21790,VINTAGE SNAP CARDS,9,2010-12-01 14:32:00,1.66,NaN,United Kingdom
1448,536544,21791,VINTAGE HEADS AND TAILS CARD GAME,2,2010-12-01 14:32:00,2.51,NaN,United Kingdom
1449,536544,21801,CHRISTMAS TREE DECORATION WITH BELL,10,2010-12-01 14:32:00,0.43,NaN,United Kingdom
1450,536544,21802,CHRISTMAS TREE HEART DECORATION,9,2010-12-01 14:32:00,0.43,NaN,United Kingdom
1451,536544,21803,CHRISTMAS TREE STAR DECORATION,11,2010-12-01 14:32:00,0.43,NaN,United Kingdom


In [41]:
rawData["CustomerID"].isnull().sum()

np.int64(135080)

CustomerID — Missing Values
Issue: There are 135,080 missing values in the CustomerID column, which is 24.93% of the dataset.

Action: I kept these values as missing (NaN) instead of filling or deleting them.

Why: CustomerID is needed to identify individual customers, but the other transaction details are still available. Therefore, these records can still be used for overall sales, product, and country analysis.

How handled: For customer-specific analysis, I will use only records where CustomerID is available.

Risk if not handled: Filling the missing values with an incorrect ID could create false customer information, while deleting the records would remove a large portion of useful transaction data.

2.Remove Duplicares


In [46]:
duplicate_count=rawData.duplicated().sum()
print("Duplicate percentage:", (duplicate_count / len(rawData)) * 100)

Duplicate percentage: 0.9721189350979592


In [47]:
rawData[rawData.duplicated(keep=False)].head(20)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
485,536409,22111,SCOTTIE DOG HOT WATER BOTTLE,1,2010-12-01 11:45:00,4.95,17908.0,United Kingdom
489,536409,22866,HAND WARMER SCOTTY DOG DESIGN,1,2010-12-01 11:45:00,2.10,17908.0,United Kingdom
494,536409,21866,UNION JACK FLAG LUGGAGE TAG,1,2010-12-01 11:45:00,1.25,17908.0,United Kingdom
517,536409,21866,UNION JACK FLAG LUGGAGE TAG,1,2010-12-01 11:45:00,1.25,17908.0,United Kingdom
521,536409,22900,SET 2 TEA TOWELS I LOVE LONDON,1,2010-12-01 11:45:00,2.95,17908.0,United Kingdom
527,536409,22866,HAND WARMER SCOTTY DOG DESIGN,1,2010-12-01 11:45:00,2.10,17908.0,United Kingdom
537,536409,22900,SET 2 TEA TOWELS I LOVE LONDON,1,2010-12-01 11:45:00,2.95,17908.0,United Kingdom
539,536409,22111,SCOTTIE DOG HOT WATER BOTTLE,1,2010-12-01 11:45:00,4.95,17908.0,United Kingdom
548,536412,22327,ROUND SNACK BOXES SET OF 4 SKULLS,1,2010-12-01 11:49:00,2.95,17920.0,United Kingdom
555,536412,22327,ROUND SNACK BOXES SET OF 4 SKULLS,1,2010-12-01 11:49:00,2.95,17920.0,United Kingdom


In [48]:
for col in rawData.columns:
    print(f"\n--- {col} ---")
    print("Duplicate values:", rawData[col].duplicated().sum())


--- InvoiceNo ---
Duplicate values: 516009

--- StockCode ---
Duplicate values: 537839

--- Description ---
Duplicate values: 537685

--- Quantity ---
Duplicate values: 541187

--- InvoiceDate ---
Duplicate values: 518649

--- UnitPrice ---
Duplicate values: 540279

--- CustomerID ---
Duplicate values: 537536

--- Country ---
Duplicate values: 541871


I distinguished between repeated values and duplicate records. Repeated values in individual columns are normal in transaction data. For example, one invoice can contain multiple products and one customer can make multiple purchases. Therefore, I only considered removing exact duplicate rows where the complete record was identical.

3.Correcting Data Types

In [50]:
rawData.dtypes

InvoiceNo              object
StockCode              object
Description            object
Quantity                int64
InvoiceDate    datetime64[ns]
UnitPrice             float64
CustomerID            float64
Country                object
dtype: object

Data Type Validation: I checked the data types of all columns. The existing data types were appropriate for their respective fields, so no major data type conversion was required. InvoiceDate was already in datetime format, while Quantity and UnitPrice were correctly stored as numeric values. CustomerID was retained as a numeric field because it contains missing values.

4.Standardizing data 


In [51]:
rawData["Country"].unique()

array(['United Kingdom', 'France', 'Australia', 'Netherlands', 'Germany',
       'Norway', 'EIRE', 'Switzerland', 'Spain', 'Poland', 'Portugal',
       'Italy', 'Belgium', 'Lithuania', 'Japan', 'Iceland',
       'Channel Islands', 'Denmark', 'Cyprus', 'Sweden', 'Austria',
       'Israel', 'Finland', 'Bahrain', 'Greece', 'Hong Kong', 'Singapore',
       'Lebanon', 'United Arab Emirates', 'Saudi Arabia',
       'Czech Republic', 'Canada', 'Unspecified', 'Brazil', 'USA',
       'European Community', 'Malta', 'RSA'], dtype=object)

In [52]:
rawData["Description"].str.strip().ne(rawData["Description"]).sum()

np.int64(113793)

In [53]:
description_spaces = rawData[
    rawData["Description"].notna() &
    rawData["Description"].str.strip().ne(rawData["Description"])
]

print("Descriptions with extra spaces:", len(description_spaces))

Descriptions with extra spaces: 113681


In [54]:
description_spaces[["Description"]].head(20)

,Description
10,POPPY'S PLAYHOUSE BEDROOM
13,IVORY KNITTED MUG COSY
15,BOX OF VINTAGE JIGSAW BLOCKS
27,ALARM CLOCK BAKELIKE RED
30,STARS GIFT TAPE
31,INFLATABLE POLITICAL GLOBE
32,VINTAGE HEADS AND TAILS CARD GAME
33,SET/2 RED RETROSPOT TEA TOWELS
34,ROUND SNACK BOXES SET OF4 WOODLAND
35,SPACEBOY LUNCH BOX


In [ ]:
rawData["Quantity"].head(20)
rawData["InvoiceDate"].head(20)
rawData["StockCode"].head(20)
rawData["CustomerID"].head(20)
rawData["UnitPrice"].head(20)
rawData["InvoiceNo"].head(20)



5.Creating calculated fields 


In [71]:
rawData["Revenue"] = rawData["Quantity"] * rawData["UnitPrice"] 
rawData["CancellationFlag"] = rawData["InvoiceNo"].astype(str).str.startswith("C")
rawData["Month"] = rawData["InvoiceDate"].dt.to_period("M").astype(str)
rawData["Year"] = rawData["InvoiceDate"].dt.year
rawData.dtypes


InvoiceNo                   object
StockCode                   object
Description                 object
Quantity                     int64
InvoiceDate         datetime64[ns]
UnitPrice                  float64
CustomerID                 float64
Country                     object
Revenue                    float64
CancellationFlag              bool
Month                       object
Year                         int32
dtype: object

In [73]:
rawData.head(10)


,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,Revenue,CancellationFlag,Month,Year
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom,15.30,False,2010-12,2010
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom,20.34,False,2010-12,2010
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom,22.00,False,2010-12,2010
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom,20.34,False,2010-12,2010
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom,20.34,False,2010-12,2010
5,536365,22752,SET 7 BABUSHKA NESTING BOXES,2,2010-12-01 08:26:00,7.65,17850.0,United Kingdom,15.30,False,2010-12,2010
6,536365,21730,GLASS STAR FROSTED T-LIGHT HOLDER,6,2010-12-01 08:26:00,4.25,17850.0,United Kingdom,25.50,False,2010-12,2010
7,536366,22633,HAND WARMER UNION JACK,6,2010-12-01 08:28:00,1.85,17850.0,United Kingdom,11.10,False,2010-12,2010
8,536366,22632,HAND WARMER RED POLKA DOT,6,2010-12-01 08:28:00,1.85,17850.0,United Kingdom,11.10,False,2010-12,2010
9,536367,84879,ASSORTED COLOUR BIRD ORNAMENT,32,2010-12-01 08:34:00,1.69,13047.0,United Kingdom,54.08,False,2010-12,2010


In [76]:
negative_quantity = (rawData["Quantity"] < 0).sum()

print("Negative Quantity rows:", negative_quantity)

Negative Quantity rows: 10624


5.Categorizing data 


In [77]:
rawData["SalesType"] = np.where(
    rawData["CancellationFlag"] == True,
    "Cancellation/Return",
    "Sale"
)

In [79]:
rawData.head(200)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,Revenue,CancellationFlag,Month,Year,SalesType
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom,15.30,False,2010-12,2010,Sale
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom,20.34,False,2010-12,2010,Sale
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom,22.00,False,2010-12,2010,Sale
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom,20.34,False,2010-12,2010,Sale
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom,20.34,False,2010-12,2010,Sale
...,...,...,...,...,...,...,...,...,...,...,...,...,...
195,536388,22469,HEART OF WICKER SMALL,12,2010-12-01 09:59:00,1.65,16250.0,United Kingdom,19.80,False,2010-12,2010,Sale
196,536388,22242,5 HOOK HANGER MAGIC TOADSTOOL,12,2010-12-01 09:59:00,1.65,16250.0,United Kingdom,19.80,False,2010-12,2010,Sale
197,536389,22941,CHRISTMAS LIGHTS 10 REINDEER,6,2010-12-01 10:03:00,8.50,12431.0,Australia,51.00,False,2010-12,2010,Sale
198,536389,21622,VINTAGE UNION JACK CUSHION COVER,8,2010-12-01 10:03:00,4.95,12431.0,Australia,39.60,False,2010-12,2010,Sale


Outliers

In [80]:
Q1 = rawData["UnitPrice"].quantile(0.25)
Q3 = rawData["UnitPrice"].quantile(0.75)

IQR = Q3 - Q1

lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

price_outliers = rawData[
    (rawData["UnitPrice"] < lower_bound) |
    (rawData["UnitPrice"] > upper_bound)
]

print("Lower bound:", lower_bound)
print("Upper bound:", upper_bound)
print("UnitPrice outliers:", len(price_outliers))

Lower bound: -3.0700000000000003
Upper bound: 8.45
UnitPrice outliers: 39627


In [81]:
rawData[["StockCode", "Description", "Quantity", "UnitPrice", "Revenue"]].sort_values(
    "UnitPrice", ascending=False
).head(20)

,StockCode,Description,Quantity,UnitPrice,Revenue
222681,M,Manual,-1,38970.00,-38970.00
524602,AMAZONFEE,AMAZON FEE,-1,17836.46,-17836.46
43702,AMAZONFEE,AMAZON FEE,-1,16888.02,-16888.02
43703,AMAZONFEE,AMAZON FEE,-1,16453.71,-16453.71
15016,AMAZONFEE,AMAZON FEE,-1,13541.33,-13541.33
15017,AMAZONFEE,AMAZON FEE,1,13541.33,13541.33
16356,AMAZONFEE,AMAZON FEE,-1,13541.33,-13541.33
16232,AMAZONFEE,AMAZON FEE,-1,13474.79,-13474.79
524601,AMAZONFEE,AMAZON FEE,-1,11586.50,-11586.50
299982,B,Adjust bad debt,1,11062.06,11062.06


Insights


In [82]:
# 1. Monthly revenue
monthlyRevenue = (
    rawData.groupby("Month")["Revenue"]
    .sum()
    .sort_values(ascending=False)
)

# 2. Top products by revenue
topProducts = (
    rawData.groupby("Description")["Revenue"]
    .sum()
    .sort_values(ascending=False)
    .head(10)
)

# 3. Top customers by revenue
topCustomers = (
    rawData.dropna(subset=["CustomerID"])
    .groupby("CustomerID")["Revenue"]
    .sum()
    .sort_values(ascending=False)
    .head(10)
)

# 4. Top countries by revenue
topCountries = (
    rawData.groupby("Country")["Revenue"]
    .sum()
    .sort_values(ascending=False)
    .head(10)
)

# 5. Cancellation analysis
cancellationSummary = rawData.groupby("SalesType").agg(
    Transactions=("InvoiceNo", "count"),
    Revenue=("Revenue", "sum")
)

print("===== TOP MONTHS =====")
print(monthlyRevenue.head(5))

print("\n===== TOP PRODUCTS =====")
print(topProducts)

print("\n===== TOP CUSTOMERS =====")
print(topCustomers)

print("\n===== TOP COUNTRIES =====")
print(topCountries)

print("\n===== CANCELLATION SUMMARY =====")
print(cancellationSummary)

===== TOP MONTHS =====
Month
2011-11    1461756.250
2011-10    1070704.670
2011-09    1019687.622
2010-12     748957.020
2011-05     723333.510
Name: Revenue, dtype: float64

===== TOP PRODUCTS =====
Description
DOTCOM POSTAGE                        206245.48
REGENCY CAKESTAND 3 TIER              164762.19
WHITE HANGING HEART T-LIGHT HOLDER     99668.47
PARTY BUNTING                          98302.98
JUMBO BAG RED RETROSPOT                92356.03
RABBIT NIGHT LIGHT                     66756.59
POSTAGE                                66230.64
PAPER CHAIN KIT 50'S CHRISTMAS         63791.94
ASSORTED COLOUR BIRD ORNAMENT          58959.73
CHILLI LIGHTS                          53768.06
Name: Revenue, dtype: float64

===== TOP CUSTOMERS =====
CustomerID
14646.0    279489.02
18102.0    256438.49
17450.0    187482.17
14911.0    132572.62
12415.0    123725.45
14156.0    113384.14
17511.0     88125.38
16684.0     65892.08
13694.0     62653.10
15311.0     59419.34
Name: Revenue, dtype: float64


In [83]:
rawData.to_excel("processedData.xlsx", index=False)